### DISCUSSION TITLE
Differentiating morph recognition from morph production.

### DISCUSSION THREAD

**Discussion by ashbywinch** [mortii/anki-morphs#279] [Ideas]
### Is your feature request related to a problem?

My cards are half "recognition", so a phrase in the target language to recognise, and half "production", a phrase in the source language to translate into the target language.
But, whether I "know" a word or not depends completely on whether we're talking recognition or production. For most words, I'll know them to recognise well before I know them to produce.

This means that Ankimorphs serves me up lots of production cards that it thinks have only one unknown morph, but which in fact are very difficult and have many morphs I can't reliably produce yet (even though I recognise them).

Speak in a new language seems like the hardest part of language learning, so it would be great if Anki could help out with it even more effectively.

### Describe the solution you would like

In the "Note filters", it would be possible to add two extra columns, one column for "card" (which card of the note, from the Card field) (could default to "all") and one column for "recognition/production" (maybe defaulting to "recognition", because if I can produce a word I will definitely recognise it). Then the "known morphs/lemmas" could be calculated separately for "morphs/lemmas I recognise" and "morphs/lemmas I can produce". And be applied only to the relevant cards.

I suggest that the figures on the toolbar thing be reduced to only morphs OR lemmas and presented something like
📖 1000    📢 500

### Describe alternatives you've considered

Currently I resolve this by throttling all my "production" decks - reducing "new cards" on those specific decks, so I don't get lots of too-hard cards every day. But that doesn't resolve the underlying problem that the cards are too hard. And I'd prefer to get more correctly-levelled "production" cards so my speaking could improve.

### Additional context

Ankimorphs is fantastic and has hugely improved my Anki experience - thanks for all the hard work.

**Comment by mortii** [comment]
Thanks for the suggestion, it's definitely an interesting problem.

> Then the "known morphs/lemmas" could be calculated separately for "morphs/lemmas I recognise" and "morphs/lemmas I can produce". And be applied only to the relevant cards.

Completely separate? So if I know 1k morphs from recognition cards, would I start over from 0 when I start using production cards?

**Reply by ashbywinch** [reply]
Yup - I don't see a better way to estimate it. I figure most people are doing a mix of the both kinds of cards from the start, so it would be unusual to get in a situation where you have lots of known morphs from recognition, but none from production. If someone did have that problem, they could set all their Ankimorph filters to ignore the distinction completely (don't set any filters to "production").

On the other hand, though - I think it's fair to assume that "morphs I can recognise" is a superset of "morphs I can produce". So if I know 1k morphs from production cards I don't need to start from 0 when using recognition cards.

**Comment by ashbywinch** [comment]
I just realised another use case for this exact feature (the "Card" filter field, not the recognition/production field):

I have a deck where each note is for one word, and then it has an example sentence field. There are four card types: two for the single word, and then two more for the example sentence. Right now, I configure the Ankimorph filter for this note type to use the example sentence. But then, if I have a single-word-card with a word that I should really learn tomorrow, but the first example sentence is long and weird, then that card will get sorted to the end of the list. Or the other way round, I might see the word card unnecessarily because it's being scored the same as the really great example sentence.
 
In general, I don't think it's always correct for all the cards of one note to have the same score.

**Comment by mortii** [comment]
I'm concerned that the proposed changes would exacerbate the following:
1. The add-on is already too complicated for most users.
2. The note filters are already too wide.

> In general, I don't think it's always correct for all the cards of one note to have the same score.

I agree. Adding a `card type` column is a clever solution to the sibling cards issue, which has come up a few times. Even though it clashes with my concerns above, I’m open to giving it a try.

> I figure most people are doing a mix of the both kinds of cards from the start

I don't have any numbers on this, but you are the first one to raise this as an issue, which suggests the opposite. Unfortunately, I think your original suggestions would pivot the add-on too hard into this recognition-production paradigm.

One potential workaround for your production issue is to create a separate Anki profile just for those cards. This would give you the disconnected behavior you’re looking for, though the downside is that this second profile won’t sync with AnkiWeb.

**Reply by aleksejrs** [reply]
> I don't have any numbers on this, but you are the first one to raise this as an issue, which suggests the opposite.

I think I raised it first. That's why I switched to FrequencyMan.

**Reply by mortii** [reply]
>I think I raised it first. That's why I switched to FrequencyMan.

Ah, yes, I remember now. HQYang also mentioned it I believe, so it's at least three people, my bad.

**Reply by ashbywinch** [reply]
Aha, I think "switch to FrequencyMan" might be the solution I'm looking for! "Too complicated for most users" is fine by me :-)
Thanks for the suggestion @aleksejrs

**Reply by sunyatasattva** [reply]
I would just like to add my voice here. I just switched from Morphman to AnkiMorph and it's a huge upgrade! Thank you for this amazing work!

I have the exact same use-case as @ashbywinch: one card for recognition and one for production. So, I guess it makes at least four of us. 🙃 

As for the complexity aspect, I've got a few comments:

1. I think people attracted to this add-on are already sort of Anki “power-users”.
2. Complexity issues can usually be (at least partially) solved by UX and UI design. And by setting defaults which work for most people.

I'm a fellow dev and also have a background in UI/UX design, so—if need be—I'm willing to help (though I have no experience with developing for Anki).

**Reply by mortii** [reply]
@sunyatasattva I (obviously) have minimal experience with UI/UX so any help/suggestions would definitely be very welcome! 

I **really** dislike the look the note filters:

![Screenshot 2024-09-26 113950](https://github.com/user-attachments/assets/677ed2c5-0144-4fea-bff9-219bf6654e0f)

Adding one more column might be fine, but the width already makes them unwieldy imo. Can you think of any better layouts, or do you think it's fine the way it is?

**Reply by sunyatasattva** [reply]
Hey @mortii! Thanks for taking on my offer, and sorry for having taken a while to reply. I dedicated some time to rethink about the way settings could be visualized. I think that the main reason you didn't like the current UI was because it got fixed in one way of thinking about the data: I understand that as developers, it's easy to model the UI closely after the underlying data structures, like SQL tables. While this makes sense from a technical standpoint, it can sometimes lead to interfaces that feel rigid or “unwieldy” as you described.

I have explored two alternatives, but I really dedicated more time to one which seemed the best way forward to me.

## Collapsible cards

<img width="785" alt="Screenshot 2024-10-01 at 22 57 27" src="https://github.com/user-attachments/assets/f64b26e9-e7e0-4d11-a86c-b260a515ead9">

The idea is to replace the table format with individual, collapsible cards for each note filter. This breaks the settings into manageable sections, reducing visual clutter and making it easier to focus on one filter at a time. Important elements like note types are more prominent, and settings are logically grouped within each card.

Each card can be expanded to reveal more details or collapsed to hide them. This allows users to control the amount of information they see at any given time.

I also changed the checkboxes to more intuitive controls, at least based on my understanding of what they do.

1. As far as I understand, the “Read” checkbox is basically just whether the filter should be respected or not (i.e. is active). So I changed that into a toggle.
2. The “Modify” checkbox becomes a “chip” on the header of the card.

Ideally, labels which are hard to understand will have tooltips on hover. Even though one can understand more by reading the AnkiMorph docs, there is no reason IMO to not help the users there. Especially when we can do so at no cost of space or readability.

I changed the tags to be two fields, one for includes and one for excludes, and I used a “token”-style text input, which is more readable and user-friendly than writing JSON by hand.

###  Advantages over the previous UI

1. **Improved Visual Hierarchy and Readability:** The selected note type is now more prominent in each card's header, helping users quickly identify and differentiate between filters. Since it's going to be common to have multiple filters with the same note type but different tags, tags are also prominently shown in the collapsed view.
2. **Reduced Cognitive Load:** By moving away from a dense, table-based layout, we minimize the feeling of being overwhelmed. 
3. **Flexibility in Data Presentation:** We're no longer constrained by the technical model of rows and columns. Adding more options to the filters (such as the one suggested in this thread) should now be more intuitive and manageable, both from a technical perspective, and from the user.

## Sidebar view
<img width="889" alt="Screenshot 2024-10-01 at 22 54 54" src="https://github.com/user-attachments/assets/81270ac4-2b64-4636-9976-bcd98a5f95fc">

I briefly explored another possibility, to have note types on the side bar and a “detail” view on the right side, mimicking the UI of, e.g., email clients.

I haven't explored this paradigm as much (as you can see), but I wanted to throw it out there for inspiration, and to show that I have considered other alternatives. Ultimately, I think I like the collapsible card based approach best.

---

Once again, I don't know much about Anki development, and I don't know how hard would this be to implement using QT. However, I hope I provided some inspiration for how the next steps could look like. I didn't add anything about the cards and the “production/recognition” suggested here in this thread on purpose.

This is a starting point for next steps in the design, and depending on your intentions we can figure out how to add those suggested features.

Hope it is somewhat helpful!

**Reply by mortii** [reply]
@sunyatasattva wow, that is absolutely incredible, thank you so much! The sidebar view approach would be a lot easier to implement, but I really like the look of the collapsible cards, so I'll try that first. Thanks again! 🙏

**Comment by mortii** [comment]
[Not adding any new features](https://github.com/mortii/anki-morphs/releases/tag/v3.3.0), apologies.

**Reply by sunyatasattva** [reply]
@mortii Aah, sad to hear that. Loved the add-on and its potential. Cheers! 🌟
